scipy/linalg/decomp_qr.py

```diff
@@ -393,24 +387,15 @@ def rq(a, overwrite_a=False, lwork=None, mode='full', check_finite=True):
     gor_un_grq, = get_lapack_funcs(('orgrq',), (rq,))
 
     if N < M:
-        # get optimal work array
-        Q, work, info = gor_un_grq(rq[-N:], tau, lwork=-1, overwrite_a=1)
-        lwork = work[0].real.astype(numpy.int)
-        Q, work, info = gor_un_grq(rq[-N:], tau, lwork=lwork, overwrite_a=1)
+        Q, = safecall(gor_un_grq, "gorgrq/gungrq", rq[-N:], tau, lwork=lwork,
```

Comment:
I see what you mean, but I think this approach will work, because `lwork` is never overridden. So if `lwork=None` in the `rq` function call, then for both the `gerqf` and `gor_un_grq` safecall, the `lwork` parameter will be None and the optimal size would be computed first.

This is also the way it works in the existing `qr` function.

It appears that both [orgrq](https://software.intel.com/en-us/node/521025) and [gerqf](https://software.intel.com/en-us/node/521024) have the same requirements and optimal `lwork`, so I think sharing across all of them isn't a problem in the case that `lwork` is explicitly set to something other than `None` or `-1`.